### PR TITLE
Use globally installed package managers

### PR DESCRIPTION
**Pull request by Turbo87**
[Rendered](https://github.com/Turbo87/ember-cli-rfcs/blob/use-global-package-managers/active/0000-use-global-package-managers.md)

### CODE REVIEW

**Comment by webark** [comment]
another potential alternative could be adding them to the package.json blueprint rather then the ember-cli package itself. That enables a kind of "opt in/out" fit the user, as well as informing what version ideally would be used. Ember init might also be able to detect which version you have, and optionally include it as well. This might be harder to teach though.

**Comment by stefanpenner** [comment]
Yes not apposed to re-evaluating this as the landscape has changed. We could also lazily install the version we need, if the version provided appears invalid or incompatible.

**Comment by nathanhammond** [comment]
I'm in favor of this. I feel like it will make life much better. There will be teething pains but I think we can get it all sorted out.

**Comment by nathanhammond** [comment]
> auto-detecting `yarn` vs. `npm`

I'm opposed to doing this without explicit opt in. I have both, I use both in different scenarios. Simply having them installed doesn't communicate which was used for setting up the current project or any available links. I'd propose an addition to `.ember-cli` as well as a prompt on `ember init` if `yarn` is present.

**Comment by nathanhammond** [comment]
Capturing from Slack:

> We should likely detect if `bower` is missing then install it during `addBowerPackageToProject`.

Because `ember install some-addon` may otherwise break.

**Comment by stefanpenner** [comment]
@Turbo87 there are some constraints we should meet, but I am fine with a path to de-bundle:

we need to (at-least):

* ensure they are using a known to work version (similarly to how we warn \w node)
* Also invoking via the JS api, prevented many sub-process annoyances with node. It may be possible to mitigate these, but will require care.

**Comment by kellyselden** [comment]
What would we do if we detect system npm is buggy 2.0 version.
 * install npm 3
 * warn
 * not work

If we go with warn, we should be very "verbose" to try and sway a person to install a newer npm BEFORE they file a bug report on ember-cli, otherwise we're back to where we were before.

**Comment by stefanpenner** [comment]
@kellyselden we could do what is appropriate for a given command:

* `ember s` likely doesn't care (no change)
* `ember install` error if the wrong version is installed
* `ember addon` error if the wrong version is installed
* `ember new` error if the wrong version is installed
* etc etc

TL;DR it seems we can *nearly* lazily detect, just before we invoke `npm` et al, if that they are right before.

Although, it is likely annoying to not fail as fast as possible...

**Comment by nathanhammond** [comment]
One caveat: if we do this I feel like we need a way to dynamically update the blacklist of non-working `npm`, `yarn`, and `bower` versions remotely.

- Create an `ember-package-manager-blacklist` node module? That will get out of date if people sit on the old version installed in `node_modules` or don't `yarn upgrade`.
- Some JSON file on some server that we download? Always up-to-date, but "phones home."

**Comment by nathanhammond** [comment]
Proposed modification:

> On first `npm` or `bower` invocation if the present system version is not valid (e.g. `npm@2` or less) or non-existent (e.g. no `bower`) we trigger local installation of the desired version. This also sets us up nicely for `yarn` adoption.

And then dropping discussion of `yarn` from this RFC as we don't intend to address that here. (Maybe save off those sections for a future "adopt Yarn" RFC.)

This addresses all of my open concerns.

**Comment by rwjblue** [comment]
If we do this:

> On first npm or bower invocation if the present system version is not valid (e.g. npm@2 or less) or non-existent (e.g. no bower) we trigger local installation of the desired version.

There is no public API surface change here at all, and this is just an internal refactor. I'd be in favor of this refactoring.

---

That being said, it seems silly to make any sort of public facing change in this space without addressing the yarn migration.  If we decide to move forward with the internal refactorings (essentially closing this RFC as not public API) we will still need to RFC to handle `yarn` support...

**Comment by nathanhammond** [comment]
> essentially closing this RFC as not public API

This is true, we can make the refactoring changes here for just-in-time installation without an RFC. We're _very_ close to being able to do that.

**Comment by rwjblue** [comment]
FYI - `bower` has already been swapped to just in time in https://github.com/ember-cli/ember-cli/pull/6631.

**Comment by rwjblue** [comment]
I think this RFC can be closed (since this is a good internal refactoring, not changing public API, and doesn't need a bunch of internal API design).

@ember-cli/core - Thoughts?

**Comment by Turbo87** [comment]
tl;dr from this weeks team call is as @rwjblue suggested in his last comment. we consider this an internal refactoring and will move forward in using the global NPM if it is v3 or above, on v2 we provide a helpful message telling you how to upgrade your local NPM.
